## Importamos las librerias necesarias

In [24]:
import pandas as pd

## Leemos el fichero csv de Git

In [25]:
url="https://raw.githubusercontent.com/Violainedlst/proyecto-da-promo-H-modulo-3-team-2-People-pulse/main/CSVs/HR_RAW_DATA.csv"
data = pd.read_csv(url, index_col=0)

pd.set_option('display.max_columns', None)

## Sacamos la información de csv para ver los tipos de datos de cada columna

In [26]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1614 entries, 0 to 1613
Data columns (total 41 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1614 non-null   object 
 1   Attrition                 1614 non-null   object 
 2   BusinessTravel            842 non-null    object 
 3   DailyRate                 1614 non-null   object 
 4   Department                302 non-null    object 
 5   DistanceFromHome          1614 non-null   int64  
 6   Education                 1614 non-null   int64  
 7   EducationField            869 non-null    object 
 8   employeecount             1614 non-null   int64  
 9   employeenumber            1183 non-null   object 
 10  EnvironmentSatisfaction   1614 non-null   int64  
 11  Gender                    1614 non-null   int64  
 12  HourlyRate                1614 non-null   object 
 13  JobInvolvement            1614 non-null   int64  
 14  JobLevel

## Recorremos las columnas para ver la información que contiene cada una

In [27]:
for column in data.columns:
  
    print(f"Columna: {column}")
    print(data[column].unique())

    print("\n--------------------\n")

Columna: Age
['51' '52' '42' '47' '46' '48' '59' '41' '56' '38' '55' '40' '58' '35'
 '45' '33' '36' '34' 'forty-seven' '53' '43' '60' '32' '37' '49' '39' '50'
 '44' '30' 'fifty-eight' '29' '31' '54' '57' '27' 'thirty-six' '28' '26'
 'fifty-five' '25' 'fifty-two' 'thirty-one' '24' 'thirty' '23' '22' '21'
 '20' 'twenty-six' '19' 'thirty-seven' '18' 'thirty-two' 'twenty-four']

--------------------

Columna: Attrition
['No' 'Yes']

--------------------

Columna: BusinessTravel
[nan 'travel_rarely' 'travel_frequently' 'non-travel']

--------------------

Columna: DailyRate
['684,0$' '699,0$' '532,0$' '359,0$' '1319,0$' '117,0$' '1435,0$' '635,0$'
 '1276,0$' '840,0$' '247,0$' '1369,0$' '201,0$' '1360,0$' '692,0$'
 '1398,0$' '286,0$' '1402,0$' '819,0$' '884,0$' '1238,0$' '515,0$'
 '1223,0$' '202,0$' '928,0$' '607,0$' '266,0$' '429,0$' '589,0$' 'nan$'
 '1180,0$' '1282,0$' '776,0$' '665,0$' '526,0$' '1034,0$' '1403,0$'
 '1499,0$' '580,0$' '859,0$' '263,0$' '1376,0$' '885,0$' '1003,0$'
 '1321,0

## Empezamos a limpiar los datos por columnas


In [28]:
# Convertimos todos los nombres de las columnas a minúsculas
data.columns = [column.lower() for column in data.columns]

# Modificamos los nombres de las columnas para mejor comprensión lectora
data.rename(columns={
    'age': 'age',
    'attrition': 'attrition',
    'businesstravel': 'business_travel',
    'dailyrate': 'daily_rate',
    'department': 'department',
    'distancefromhome': 'distance_from_home',
    'education': 'education',
    'educationfield': 'education_field',
    'employeecount': 'employee_count',
    'employeenumber': 'employee_number',
    'environmentsatisfaction': 'environment_satisfaction',
    'gender': 'gender',
    'hourlyrate': 'hourly_rate',
    'jobinvolvement': 'job_involvement',
    'joblevel': 'job_level',
    'jobrole': 'job_role',
    'jobsatisfaction': 'job_satisfaction',
    'maritalstatus': 'marital_status',
    'monthlyincome': 'monthly_income',
    'monthlyrate': 'monthly_rate',
    'numcompaniesworked': 'num_companies_worked',
    'over18': 'over_18',
    'overtime': 'overtime',
    'percentsalaryhike': 'percent_salary_hike',
    'performancerating': 'performance_rating',
    'relationshipsatisfaction': 'relationship_satisfaction',
    'standardhours': 'standard_hours',
    'stockoptionlevel': 'stock_option_level',
    'totalworkingyears': 'total_working_years',
    'trainingtimeslastyear': 'training_times_last_year',
    'worklifebalance': 'work_life_balance',
    'yearsatcompany': 'years_at_company',
    'yearsincurrentrole': 'years_in_current_role',
    'yearssincelastpromotion': 'years_since_last_promotion',
    'yearswithcurrmanager': 'years_with_curr_manager',
    'sameasmonthlyincome': 'same_as_monthly_income',
    'datebirth': 'year_birth',
    'salary': 'salary',
    'roledepartament': 'role_department',
    'numberchildren': 'number_children',
    'remotework': 'remote_work'
}, inplace=True)

In [29]:
data

age attrition business_travel daily_rate                department  \
0     51        No             NaN     684,0$                       NaN   
1     52        No             NaN     699,0$                       NaN   
2     42        No   travel_rarely     532,0$   Research & Development    
3     47        No   travel_rarely     359,0$                       NaN   
4     46        No             NaN    1319,0$                       NaN   
...   ..       ...             ...        ...                       ...   
1609  36       Yes   travel_rarely     530,0$                       NaN   
1610  45        No      non-travel     805,0$                       NaN   
1611  39        No   travel_rarely     903,0$                       NaN   
1612  36        No      non-travel    1229,0$                       NaN   
1613  46        No             NaN     566,0$                       NaN   

      distance_from_home  education   education_field  employee_count  \
0                      6          3               NaN               1   
1                      1          4     Life Sciences               1   
2                      4          2  Technical Degree               1   
3                      2          4           Medical               1   
4                      3          3  Technical Degree               1   
...                  ...        ...               ...             ...   
1609                   3          1     Life Sciences               1   
1610                   4          2               NaN               1   
1611                 -13          5               NaN               1   
1612                   8          4  Technical Degree               1   
1613                   7          2           Medical               1   

     employee_number  environment_satisfaction  gender hourly_rate  \
0              162,0                         1       0          51   
1              259,0                         3       0          65   
2              319,0                         3       0          58   
3                NaN                         1       1          82   
4                NaN                         1       1          45   
...              ...                       ...     ...         ...   
1609           967,0                         3       0          51   
1610           972,0                         3       0          57   
1611             NaN                        13       0          41   
1612           990,0                         1       0          84   
1613             NaN                         4       0          75   

      job_involvement  job_level                  job_role  job_satisfaction  \
0                   3          5        resEArch DIREcToR                  3   
1                   2          5                  ManAGeR                  3   
2                   3          5                  ManaGER                  4   
3                   3          4        ReseArCH DIrECtOr                  3   
4                   4          4          sAleS EXECUtIve                  1   
...               ...        ...                       ...               ...   
1609                2          3          saLEs ExeCUTiVe                  4   
1610                3          2    LAboRaTOry tECHNiCIAn                  2   
1611                4          3          sAlES ExECUTivE                  3   
1612                3          2          SaLes ExecUtIVe                  4   
1613                3          3   mAnUfactURInG DiRECTOr                  3   

     marital_status monthly_income  monthly_rate  num_companies_worked  \
0               NaN        19537,0          6462                     7   
1               NaN        19999,0          5678                     0   
2           Married        19232,0          4933                     1   
3           Married        17169,0         26703                     3   
4          Divorced            NaN          7739       

In [30]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1614 entries, 0 to 1613
Data columns (total 41 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age                         1614 non-null   object 
 1   attrition                   1614 non-null   object 
 2   business_travel             842 non-null    object 
 3   daily_rate                  1614 non-null   object 
 4   department                  302 non-null    object 
 5   distance_from_home          1614 non-null   int64  
 6   education                   1614 non-null   int64  
 7   education_field             869 non-null    object 
 8   employee_count              1614 non-null   int64  
 9   employee_number             1183 non-null   object 
 10  environment_satisfaction    1614 non-null   int64  
 11  gender                      1614 non-null   int64  
 12  hourly_rate                 1614 non-null   object 
 13  job_involvement             1614 

## Decidimos quitar columnas que no nos sirven

In [31]:
# Eliminar las columnas 'id' y 'same_as_monthly_income'
data_noid = data.drop(columns=["same_as_monthly_income","employee_count","salary","standard_hours", "over_18",'performance_rating',"number_children","years_in_current_role"])

# Mostrar el DataFrame resultante
data_noid

age attrition business_travel daily_rate                department  \
0     51        No             NaN     684,0$                       NaN   
1     52        No             NaN     699,0$                       NaN   
2     42        No   travel_rarely     532,0$   Research & Development    
3     47        No   travel_rarely     359,0$                       NaN   
4     46        No             NaN    1319,0$                       NaN   
...   ..       ...             ...        ...                       ...   
1609  36       Yes   travel_rarely     530,0$                       NaN   
1610  45        No      non-travel     805,0$                       NaN   
1611  39        No   travel_rarely     903,0$                       NaN   
1612  36        No      non-travel    1229,0$                       NaN   
1613  46        No             NaN     566,0$                       NaN   

      distance_from_home  education   education_field employee_number  \
0                      6          3               NaN           162,0   
1                      1          4     Life Sciences           259,0   
2                      4          2  Technical Degree           319,0   
3                      2          4           Medical             NaN   
4                      3          3  Technical Degree             NaN   
...                  ...        ...               ...             ...   
1609                   3          1     Life Sciences           967,0   
1610                   4          2               NaN           972,0   
1611                 -13          5               NaN             NaN   
1612                   8          4  Technical Degree           990,0   
1613                   7          2           Medical             NaN   

      environment_satisfaction  gender hourly_rate  job_involvement  \
0                            1       0          51                3   
1                            3       0          65                2   
2                            3       0          58                3   
3                            1       1          82                3   
4                            1       1          45                4   
...                        ...     ...         ...              ...   
1609                         3       0          51                2   
1610                         3       0          57                3   
1611                        13       0          41                4   
1612                         1       0          84                3   
1613                         4       0          75                3   

      job_level                  job_role  job_satisfaction marital_status  \
0             5        resEArch DIREcToR                  3            NaN   
1             5                  ManAGeR                  3            NaN   
2             5                  ManaGER                  4        Married   
3             4        ReseArCH DIrECtOr                  3        Married   
4             4          sAleS EXECUtIve                  1       Divorced   
...         ...                       ...               ...            ...   
1609          3          saLEs ExeCUTiVe                  4        Married   
1610          2    LAboRaTOry tECHNiCIAn                  2            NaN   
1611          3          sAlES ExECUTivE                  3         Single   
1612          2          SaLes ExecUtIVe                  4       Divorced   
1613          3   mAnUfactURInG DiRECTOr                  3            NaN   

     monthly_income  monthly_rate  num_companies_worked overtime  \
0           19537,0          6462                     7       No   
1           19999,0          5678                     0      NaN   
2           19232,0          4933                     1       No   
3           17169,0         26703                     3      NaN   
4               NaN          7739                     2       No   
...             ...          

### Lo que hay que cambiar de la 0 a la 21

- daily_rate: quitar símbolo $ y convertir a int

In [32]:
# Aseguramos que los valores de 'daily_rate' son de tipo string
data_noid['daily_rate'] = data_noid['daily_rate'].astype(str)

# Lista de símbolos a eliminar en los valores de la columna
quitar_simbolos = [' ', '$']

# Itera sobre cada símbolo en la lista de símbolos a eliminar
for simbolo in quitar_simbolos:
    # Elimina el símbolo actual de todos los valores de la columna 'daily_rate'
    data_noid['daily_rate'] = data_noid['daily_rate'].str.replace(simbolo, '', regex=False)

# Rellena los valores NaN y los valores vacíos con '0'
data_noid['daily_rate'] = data_noid['daily_rate'].replace(['nan', ''], '0')

# Reemplazamos las comas por puntos
data_noid['daily_rate'] = data_noid['daily_rate'].str.replace(',', '.')

# Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
data_noid['daily_rate'] = pd.to_numeric(data_noid['daily_rate'], errors='coerce') 

# Cambiamos los valores de tipo string a int usando ".astype(int)"
data_noid['daily_rate'] = data_noid['daily_rate'].astype("int32")

- `age` es `object` y debería ser `int`
- `hourly_rate` aparece como `object` y debería ser `int`
- `monthly_income`: convertir a `int`

In [33]:
columnas3 = ['age', 'hourly_rate']

# Itera sobre cada columna en la lista de columnas a convertir
for columna in columnas3:
    # Asegúrate de que los valores de la columna son de tipo string
    data_noid[columna] = data_noid[columna].astype(str)
    # Elimina cualquier valor no numérico o vacío y reemplaza por 0
    data_noid[columna] = data_noid[columna].replace(['nan', '', ' '], '0')
    # Convierte la columna a tipo numérico (int)
    data_noid[columna] = pd.to_numeric(data_noid[columna], errors='coerce').fillna(0).astype("int32")

# Muestra las primeras filas del DataFrame para verificar que los valores de las columnas se han modificado correctamente
data_noid.head(2)

age attrition business_travel  daily_rate department  distance_from_home  \
0   51        No             NaN         684        NaN                   6   
1   52        No             NaN         699        NaN                   1   

   education education_field employee_number  environment_satisfaction  \
0          3             NaN           162,0                         1   
1          4   Life Sciences           259,0                         3   

   gender  hourly_rate  job_involvement  job_level             job_role  \
0       0           51                3          5   resEArch DIREcToR    
1       0           65                2          5             ManAGeR    

   job_satisfaction marital_status monthly_income  monthly_rate  \
0                 3            NaN        19537,0          6462   
1                 3            NaN        19999,0          5678   

   num_companies_worked overtime  percent_salary_hike  \
0                     7       No                   13   
1                     0      NaN                   14   

   relationship_satisfaction  stock_option_level total_working_years  \
0                          3                   0                 NaN   
1                          1                   1                34,0   

   training_times_last_year work_life_balance  years_at_company  \
0                         5               3,0                20   
1                         5               3,0                33   

   years_since_last_promotion  years_with_curr_manager  year_birth  \
0                          15                       15        1972   
1                          11                        9        1971   

  role_department remote_work  
0             NaN         Yes  
1             NaN           1

- cambiar todos los `valores` a `minúsculas`

In [34]:
# Lista de columnas a convertir a minúsculas de al 0 a la 21
columns_to_lower = ['age', 'attrition', 'business_travel', 'daily_rate', 'department',
       'distance_from_home', 'education', 'education_field', 'employee_number',
       'environment_satisfaction', 'gender', 'hourly_rate', 'job_involvement',
       'job_level', 'job_role', 'job_satisfaction', 'marital_status',
       'monthly_income', 'monthly_rate', 'num_companies_worked',
       'overtime', 'percent_salary_hike',
       'relationship_satisfaction', 'stock_option_level',
         'training_times_last_year', 'work_life_balance',
       'years_at_company', 
       'years_since_last_promotion', 'years_with_curr_manager', 'year_birth',
       'role_department', 'remote_work']

# Convertir a minúsculas las columnas especificadas
for column in columns_to_lower:
    if data_noid[column].dtype == 'object':  # Verifica si la columna es de tipo str
        data_noid[column] = data_noid[column].str.lower()
        
data_noid.head(5)


age attrition business_travel  daily_rate                department  \
0   51        no             NaN         684                       NaN   
1   52        no             NaN         699                       NaN   
2   42        no   travel_rarely         532   research & development    
3   47        no   travel_rarely         359                       NaN   
4   46        no             NaN        1319                       NaN   

   distance_from_home  education   education_field employee_number  \
0                   6          3               NaN           162,0   
1                   1          4     life sciences           259,0   
2                   4          2  technical degree           319,0   
3                   2          4           medical             NaN   
4                   3          3  technical degree             NaN   

   environment_satisfaction  gender  hourly_rate  job_involvement  job_level  \
0                         1       0           51                3          5   
1                         3       0           65                2          5   
2                         3       0           58                3          5   
3                         1       1           82                3          4   
4                         1       1           45                4          4   

              job_role  job_satisfaction marital_status monthly_income  \
0   research director                  3            NaN        19537,0   
1             manager                  3            NaN        19999,0   
2             manager                  4        married        19232,0   
3   research director                  3        married        17169,0   
4     sales executive                  1       divorced            NaN   

   monthly_rate  num_companies_worked overtime  percent_salary_hike  \
0          6462                     7       no                   13   
1          5678                     0      NaN                   14   
2          4933                     1       no                   11   
3         26703                     3      NaN                   19   
4          7739                     2       no                   12   

   relationship_satisfaction  stock_option_level total_working_years  \
0                          3                   0                 NaN   
1                          1                   1                34,0   
2                          4                   0                22,0   
3                          2                   2                 NaN   
4                          4                   1                 NaN   

   training_times_last_year work_life_balance  years_at_company  \
0                         5               3,0                20   
1                         5               3,0                33   
2                         3               NaN                22   
3                         2               NaN                20   
4                         5               3,0                19   

   years_since_last_promotion  years_with_curr_manager  year_birth  \
0                          15                       15        1972   
1                          11                        9        1971   
2                          11                       15        1981   
3                           5                        6        1976   
4                           2                        8        1977   

                        role_department remote_work  
0                                   NaN         yes  
1                                   NaN           1  
2   manager  -  research & development            1  
3                                   NaN       false  
4                                   NaN           0

- quitar los `NaN` de las columnas que lo tengan

In [35]:
# Reemplazar NaN en las columnas especificadas por 'N/A'
columns_to_replace = ['business_travel', 'department', 'education_field', 'marital_status','overtime','work_life_balance']
replacement_value = 'N/A'

data_noid[columns_to_replace] = data_noid[columns_to_replace].fillna(replacement_value)

# Mostrar los primeros registros para verificar
data_noid.head(2)


age attrition business_travel  daily_rate department  distance_from_home  \
0   51        no             N/A         684        N/A                   6   
1   52        no             N/A         699        N/A                   1   

   education education_field employee_number  environment_satisfaction  \
0          3             N/A           162,0                         1   
1          4   life sciences           259,0                         3   

   gender  hourly_rate  job_involvement  job_level             job_role  \
0       0           51                3          5   research director    
1       0           65                2          5             manager    

   job_satisfaction marital_status monthly_income  monthly_rate  \
0                 3            N/A        19537,0          6462   
1                 3            N/A        19999,0          5678   

   num_companies_worked overtime  percent_salary_hike  \
0                     7       no                   13   
1                     0      N/A                   14   

   relationship_satisfaction  stock_option_level total_working_years  \
0                          3                   0                 NaN   
1                          1                   1                34,0   

   training_times_last_year work_life_balance  years_at_company  \
0                         5               3,0                20   
1                         5               3,0                33   

   years_since_last_promotion  years_with_curr_manager  year_birth  \
0                          15                       15        1972   
1                          11                        9        1971   

  role_department remote_work  
0             NaN         yes  
1             NaN           1

In [36]:
# # Reemplazar NaN en la columna especificada por 'No'
# columns_to_replace2 = ['over_18']
# replacement_value_no = 'No'

# data_noid[columns_to_replace2] = data_noid[columns_to_replace2].fillna(replacement_value_no)

# data_noid['over_18'] = data_noid['over_18'].replace(['Y'], 'Yes')

# # Mostrar los primeros registros para verificar
# data_noid.head(2)

In [37]:
#Quitar los valores negativos y reemplazarlos por "N/A" y convertir la columna a INT
data_noid["distance_from_home"] = data_noid["distance_from_home"].apply(lambda x: "N/A" if x < 0 else x)

data_noid["distance_from_home"] = pd.to_numeric(data_noid["distance_from_home"], errors='coerce').fillna(0).astype("int32")

In [38]:
#Reemplazamos los 0 por male y 1 por female. Hemos tomado decisión de usar 0 para male y 1 para female siguiendo la logica explicada en la fase 2 del gitbook.
#Primero cambiamos la columna a tipo object para poder añadir el sexo.
data_noid["gender"] = data_noid["gender"].astype(str)

#male
data_noid["gender"] = data_noid["gender"].replace(['0'], 'male')

#female
data_noid["gender"] = data_noid["gender"].replace(['1'], 'female')

#Verificamos los datos unicos
data_noid["gender"].unique()


array(['male', 'female'], dtype=object)

In [39]:
#Reemplazamos el marreid por married dentro de la columna
data_noid["marital_status"] = data_noid["marital_status"].replace(['marreid'], 'married')

#Verificamos que esten los datos unicos
data_noid["marital_status"].unique()

array(['N/A', 'married', 'divorced', 'single'], dtype=object)

In [40]:
data_noid["monthly_income"] = data_noid["monthly_income"].fillna('0') #Reemplazar los nan por un 0 para poder luego hacer el rango

data_noid["monthly_income"] = data_noid["monthly_income"].str.replace(',', '.') #conversión de los datos cambiando las comas por punto

data_noid["monthly_income"] = data_noid["monthly_income"].astype(float) # Convertir la columna "monthly income" a tipo float

In [41]:
data_noid["monthly_income"]

0       19537.0
1       19999.0
2       19232.0
3       17169.0
4           0.0
         ...   
1609    10325.0
1610     4447.0
1611        0.0
1612        0.0
1613    10845.0
Name: monthly_income, Length: 1614, dtype: float64

In [43]:
# Convertir la columna 'employee_number' a numérico, forzando errores a NaN, y luego a entero
data['employee_number'] = pd.to_numeric(data['employee_number'].str.replace(',', ''), errors='coerce').fillna(0).astype(int)

# Primero tengo que filtrar por los números que ya existen
numeros_actuales = data[data["employee_number"] != 0]["employee_number"]

# Busco el valor máximo como principio de mi nuevo listado de números
max_numeros_actuales = numeros_actuales.max()

# Vamos a buscar cuántos empleados tienen un cero 
num_zeros = data[data["employee_number"] == 0].shape[0]

# Lista de nuevos números
nuevos_numeros = list(range(max_numeros_actuales + 1, max_numeros_actuales + 1 + num_zeros))

# Añado los números para todos los empleados que tengan un cero
data.loc[data["employee_number"] == 0, "employee_number"] = nuevos_numeros

print("DataFrame actualizado:")
data



DataFrame actualizado:


age attrition business_travel daily_rate                department  \
0     51        No             NaN     684,0$                       NaN   
1     52        No             NaN     699,0$                       NaN   
2     42        No   travel_rarely     532,0$   Research & Development    
3     47        No   travel_rarely     359,0$                       NaN   
4     46        No             NaN    1319,0$                       NaN   
...   ..       ...             ...        ...                       ...   
1609  36       Yes   travel_rarely     530,0$                       NaN   
1610  45        No      non-travel     805,0$                       NaN   
1611  39        No   travel_rarely     903,0$                       NaN   
1612  36        No      non-travel    1229,0$                       NaN   
1613  46        No             NaN     566,0$                       NaN   

      distance_from_home  education   education_field  employee_count  \
0                      6          3               NaN               1   
1                      1          4     Life Sciences               1   
2                      4          2  Technical Degree               1   
3                      2          4           Medical               1   
4                      3          3  Technical Degree               1   
...                  ...        ...               ...             ...   
1609                   3          1     Life Sciences               1   
1610                   4          2               NaN               1   
1611                 -13          5               NaN               1   
1612                   8          4  Technical Degree               1   
1613                   7          2           Medical               1   

      employee_number  environment_satisfaction  gender hourly_rate  \
0                1620                         1       0          51   
1                2590                         3       0          65   
2                3190                         3       0          58   
3               20681                         1       1          82   
4               20682                         1       1          45   
...               ...                       ...     ...         ...   
1609             9670                         3       0          51   
1610             9720                         3       0          57   
1611            21110                        13       0          41   
1612             9900                         1       0          84   
1613            21111                         4       0          75   

      job_involvement  job_level                  job_role  job_satisfaction  \
0                   3          5        resEArch DIREcToR                  3   
1                   2          5                  ManAGeR                  3   
2                   3          5                  ManaGER                  4   
3                   3          4        ReseArCH DIrECtOr                  3   
4                   4          4          sAleS EXECUtIve                  1   
...               ...        ...                       ...               ...   
1609                2          3          saLEs ExeCUTiVe                  4   
1610                3          2    LAboRaTOry tECHNiCIAn                  2   
1611                4          3          sAlES ExECUTivE                  3   
1612                3          2          SaLes ExecUtIVe                  4   
1613                3          3   mAnUfactURInG DiRECTOr                  3   

     marital_status monthly_income  monthly_rate  num_companies_worked  \
0               NaN        19537,0          6462                     7   
1               NaN        19999,0          5678                     0   
2           Married        19232,0          4933                     1   
3           Married        17169,0         26703                     3   
4          Divorced            NaN         

In [44]:
data["employee_number"]

0       162
1       259
2       319
3         0
4         0
       ... 
1609    967
1610    972
1611      0
1612    990
1613      0
Name: employee_number, Length: 1614, dtype: int32

## Limpiamos las columnas de la 21 a la 31 (ambas incluidas)

Columnas verificadas donde no había nada que modificar:

- COLUMNA 21: "num_companies_worked" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación.

- COLUMNA 24: "percent_salary_hike" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación.

- COLUMNA 26: "relationship_satisfaction" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación.

- COLUMNA 28: "stock_option_level" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación.

- COLUMNA 30: "training_times_last_year" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

- COLUMNA 32: "years_at_company" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

- COLUMNA 34: "years_since_last_promotion" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

- COLUMNA35: "years_with_curr_manager" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

- COLUMNA36: "date_birth" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

In [45]:
# Comprobación de la columna 21
data_noid["num_companies_worked"].sample(5)

1349    4
931     7
849     2
1276    4
986     8
Name: num_companies_worked, dtype: int64

In [46]:
# # LIMPIEZA DE LA COLUMNA 22 "over_18"
# nuevos_valores = {"Y": "Yes", "N": "No"}
# data_noid["over_18"] = data_noid["over_18"].map(nuevos_valores)
# data_noid['over_18'] = data_noid['over_18'].fillna('N/A') #Usamos el "fillna" para cambiar el Nan (formato numpy) a N/A

In [47]:
# Comprobación de la columna 23
data_noid["overtime"].sample(5)

336     N/A
1401    N/A
673      no
1073    N/A
484     N/A
Name: overtime, dtype: object

In [48]:
# LIMPIEZA DE LA COLUMNA 23 "overtime"
data_noid['overtime'] = data_noid['overtime'].fillna('N/A') #Usamos el "fillna" para cambiar el Nan (formato numpy) a N/A

In [49]:
# Comprobación de la columna 24
data_noid["percent_salary_hike"].sample(5)

582    13
460    14
609    14
301    18
908    15
Name: percent_salary_hike, dtype: int64

In [50]:
# Comprobación de la columna 25
# data_noid["performance_rating"].sample(5)

In [51]:
# # LIMPIEZA DE LA COLUMNA 25 "performance_rating"

# # Nos aseguramos de que todos los valores son cadenas usando .astype(str)
# data_noid["performance_rating"] = data_noid["performance_rating"].astype(str)

# # Reemplazamos las comas por puntos
# data_noid["performance_rating"] = data_noid["performance_rating"].str.replace(',', '.')

# # Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
# data_noid["performance_rating"] = pd.to_numeric(data_noid["performance_rating"], errors='coerce')

# # Cambiamos los valores que aparecen como NaN para darles valor 0
# data_noid["performance_rating"] = data_noid["performance_rating"].fillna(0) 

# # Cambiamos los valores de tipo string a int usando ".astype(int)"
# data_noid["performance_rating"] = data_noid["performance_rating"].astype(int)

In [52]:
# Comprobación de la columna 26
data_noid["relationship_satisfaction"].sample(5)

1298    3
1609    1
873     2
680     2
1441    3
Name: relationship_satisfaction, dtype: int64

In [53]:
# # LIMPIEZA DE LA COLUMNA 27 "standard_hours" COLUMNA BORRADA

# # Nos aseguramos de que todos los valores son cadenas usando .astype(str)
# data_noid["standard_hours"] = data_noid["standard_hours"].astype(str)

# # Reemplazamos las comas por puntos
# data_noid["standard_hours"] = data_noid["standard_hours"].str.replace(',', '.')

# # Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
# data_noid["standard_hours"] = pd.to_numeric(data_noid["standard_hours"], errors='coerce')

# # Cambiamos los valores que aparecen como NaN para darles valor 0
# data_noid["standard_hours"] = data_noid["standard_hours"].fillna(0) 

# # Cambiamos los valores de tipo string a int usando ".astype(int)"
# data_noid["standard_hours"] = data_noid["standard_hours"].astype(int)

In [54]:
# Comprobación de la columna 28
data_noid["stock_option_level"].sample(5)

1196    1
1423    1
439     1
25      1
74      3
Name: stock_option_level, dtype: int64

In [55]:
# # LIMPIEZA DE LA COLUMNA 29 "total_working_years"

# # Nos aseguramos de que todos los valores son cadenas usando .astype(str)
data_noid["total_working_years"] = data_noid["total_working_years"].astype(str)

# # Reemplazamos las comas por puntos
data_noid["total_working_years"] = data_noid["total_working_years"].str.replace(',', '.')

# # Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
data_noid["total_working_years"] = pd.to_numeric(data_noid["total_working_years"], errors='coerce')

# # Cambiamos los valores que aparecen como NaN para dales valor 0
data_noid["total_working_years"] = data_noid["total_working_years"].fillna(0) 

# # Cambiamos los valores de tipo string a int usando ".astype(int)"
data_noid["total_working_years"] = data_noid["total_working_years"].astype("int32")

In [56]:
# Comprobación de la columna 30
data_noid["training_times_last_year"].sample(5)

598     2
234     2
237     4
276     3
1330    6
Name: training_times_last_year, dtype: int64

In [57]:
# Comprobación de la columna 31
data_noid["work_life_balance"].sample(5)

1388    2,0
365     4,0
1491    3,0
754     3,0
1444    3,0
Name: work_life_balance, dtype: object

In [58]:
# LIMPIEZA DE LA COLUMNA 31 "work_life_balance"

# Nos aseguramos de que todos los valores son cadenas usando .astype(str)
data_noid["work_life_balance"] = data_noid["work_life_balance"].astype(str)

# Reemplazamos las comas por puntos
data_noid["work_life_balance"] = data_noid["work_life_balance"].str.replace(',', '.')

# Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
data_noid["work_life_balance"] = pd.to_numeric(data_noid["work_life_balance"], errors='coerce')

# Cambiamos los valores que aparecen como NaN para darles valor 0
data_noid["work_life_balance"] = data_noid["work_life_balance"].fillna(0) 

# Cambiamos los valores de tipo string a int usando ".astype(int)"
data_noid["work_life_balance"] = data_noid["work_life_balance"].astype("int32")

In [59]:
data_noid["work_life_balance"].unique()

array([3, 0, 2, 4, 1], dtype=int32)

In [60]:
data_noid.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1614 entries, 0 to 1613
Data columns (total 33 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age                         1614 non-null   int32  
 1   attrition                   1614 non-null   object 
 2   business_travel             1614 non-null   object 
 3   daily_rate                  1614 non-null   int32  
 4   department                  1614 non-null   object 
 5   distance_from_home          1614 non-null   int32  
 6   education                   1614 non-null   int64  
 7   education_field             1614 non-null   object 
 8   employee_number             1614 non-null   int32  
 9   environment_satisfaction    1614 non-null   int64  
 10  gender                      1614 non-null   object 
 11  hourly_rate                 1614 non-null   int32  
 12  job_involvement             1614 non-null   int64  
 13  job_level                   1614 

In [61]:
# Comprobación de la columna 32
data_noid["years_at_company"].sample(5)

1173    1
1341    0
1331    1
1138    3
431     9
Name: years_at_company, dtype: int64

In [62]:
# # Comprobación de la columna 33 COLUMNA BORRADA
# data_noid['years_in_current_role'] = data_noid['years_in_current_role'].fillna('N/A')

In [63]:
# Comprobación de la columna 34
data_noid["years_since_last_promotion"].sample(5)

1340     0
789      1
1188     0
288      1
669     11
Name: years_since_last_promotion, dtype: int64

In [64]:
# Comprobación de la columna 35
data_noid["years_with_curr_manager"].sample(5)

1075     1
661      7
31      11
715      2
1493     3
Name: years_with_curr_manager, dtype: int64

In [65]:
# Comprobación de la columna 36
data_noid["year_birth"].sample(5)

490     1981
1452    2001
248     1990
361     1982
1123    1986
Name: year_birth, dtype: int64

In [66]:
# # Comprobación de la columna 37 #Ver con equipo que quieren hacer
# data_noid["salary"].sample(25)
# data_noid["salary"] = data_noid["salary"].str.replace('1000000000$', 'Unknown')

In [67]:
# Comprobación de la columna 38

data_noid["role_department"] = data_noid["role_department"].fillna('N/A')

In [68]:
# Comprobación de la columna 39

# data_noid["number_children"] = data_noid["number_children"].fillna('N/A')

In [69]:
# Comprobación de la columna 40
data_noid["remote_work"].sample(25)

#primero paso todo a lower para usar el lambda
data_noid['remote_work'] = data_noid['remote_work'].str.lower()

# Reemplazar valores que no sean "yes" o "no" con "no"
data_noid['remote_work'] = data_noid['remote_work'].apply(lambda x: 'yes' if x == 'yes' else 'no')


In [70]:
data_noid["remote_work"].sample(25)

1214     no
304     yes
1475    yes
285      no
751      no
1361    yes
1592     no
902      no
355      no
463      no
302      no
702      no
990      no
171      no
982     yes
755      no
1189     no
1044     no
322      no
177      no
383      no
820      no
1574     no
826      no
236      no
Name: remote_work, dtype: object

In [71]:
data_noid.head(2)

age attrition business_travel  daily_rate department  distance_from_home  \
0   51        no             N/A         684        N/A                   6   
1   52        no             N/A         699        N/A                   1   

   education education_field  employee_number  environment_satisfaction  \
0          3             N/A              162                         1   
1          4   life sciences              259                         3   

  gender  hourly_rate  job_involvement  job_level             job_role  \
0   male           51                3          5   research director    
1   male           65                2          5             manager    

   job_satisfaction marital_status  monthly_income  monthly_rate  \
0                 3            N/A         19537.0          6462   
1                 3            N/A         19999.0          5678   

   num_companies_worked overtime  percent_salary_hike  \
0                     7       no                   13   
1                     0      N/A                   14   

   relationship_satisfaction  stock_option_level  total_working_years  \
0                          3                   0                    0   
1                          1                   1                   34   

   training_times_last_year  work_life_balance  years_at_company  \
0                         5                  3                20   
1                         5                  3                33   

   years_since_last_promotion  years_with_curr_manager  year_birth  \
0                          15                       15        1972   
1                          11                        9        1971   

  role_department remote_work  
0             N/A         yes  
1             N/A          no

In [72]:
for column in data_noid.columns:
        print(f"Columna: {column}")
        print(data_noid[column].unique())

        print("\n--------------------\n")               

Columna: age
[51 52 42 47 46 48 59 41 56 38 55 40 58 35 45 33 36 34  0 53 43 60 32 37
 49 39 50 44 30 29 31 54 57 27 28 26 25 24 23 22 21 20 19 18]

--------------------

Columna: attrition
['no' 'yes']

--------------------

Columna: business_travel
['N/A' 'travel_rarely' 'travel_frequently' 'non-travel']

--------------------

Columna: daily_rate
[ 684  699  532  359 1319  117 1435  635 1276  840  247 1369  201 1360
  692 1398  286 1402  819  884 1238  515 1223  202  928  607  266  429
  589    0 1180 1282  776  665  526 1034 1403 1499  580  859  263 1376
  885 1003 1321  394 1372 1333  228  737  823  667  301  573 1329  630
 1063 1017 1296  939 1355 1448  200 1202  404  208  813  465 1189 1001
 1394  161  288  682 1354  147  119 1413  452  334 1132  982  480 1099
  672 1379  583 1492 1050  469  237 1440 1291 1157 1336 1224  735 1389
  638 1240  194 1339  111 1469  470 1232 1249  798  549  570  541  164
 1117  619  319  956 1245 1397  527  213  882  330  406  217  481  669
 1465  685

In [73]:
data_noid.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1614 entries, 0 to 1613
Data columns (total 33 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age                         1614 non-null   int32  
 1   attrition                   1614 non-null   object 
 2   business_travel             1614 non-null   object 
 3   daily_rate                  1614 non-null   int32  
 4   department                  1614 non-null   object 
 5   distance_from_home          1614 non-null   int32  
 6   education                   1614 non-null   int64  
 7   education_field             1614 non-null   object 
 8   employee_number             1614 non-null   int32  
 9   environment_satisfaction    1614 non-null   int64  
 10  gender                      1614 non-null   object 
 11  hourly_rate                 1614 non-null   int32  
 12  job_involvement             1614 non-null   int64  
 13  job_level                   1614 

In [74]:
# Pasamos las columnas "int64" a "int32" para que no de errores en SQL
columnas_convertidas = ["education", "employee_number", "environment_satisfaction", "job_involvement", "job_level", "job_satisfaction", "monthly_rate", "num_companies_worked", "percent_salary_hike", "relationship_satisfaction", "stock_option_level", "training_times_last_year", "years_at_company", "years_since_last_promotion", "years_with_curr_manager", "year_birth"]

# Cambiamos los valores de tipo string a int usando ".astype(int32)"
data_noid[columnas_convertidas] = data_noid[columnas_convertidas].astype("int32")

In [75]:
data_noid.to_csv('HR_RAW_DATA_FINAL.csv')

In [76]:
for column in data_noid.columns: 
    print(f"Columna: {column}")
    print(data[column].unique())
    print("\n--------------------\n")

Columna: age
['51' '52' '42' '47' '46' '48' '59' '41' '56' '38' '55' '40' '58' '35'
 '45' '33' '36' '34' 'forty-seven' '53' '43' '60' '32' '37' '49' '39' '50'
 '44' '30' 'fifty-eight' '29' '31' '54' '57' '27' 'thirty-six' '28' '26'
 'fifty-five' '25' 'fifty-two' 'thirty-one' '24' 'thirty' '23' '22' '21'
 '20' 'twenty-six' '19' 'thirty-seven' '18' 'thirty-two' 'twenty-four']

--------------------

Columna: attrition
['No' 'Yes']

--------------------

Columna: business_travel
[nan 'travel_rarely' 'travel_frequently' 'non-travel']

--------------------

Columna: daily_rate
['684,0$' '699,0$' '532,0$' '359,0$' '1319,0$' '117,0$' '1435,0$' '635,0$'
 '1276,0$' '840,0$' '247,0$' '1369,0$' '201,0$' '1360,0$' '692,0$'
 '1398,0$' '286,0$' '1402,0$' '819,0$' '884,0$' '1238,0$' '515,0$'
 '1223,0$' '202,0$' '928,0$' '607,0$' '266,0$' '429,0$' '589,0$' 'nan$'
 '1180,0$' '1282,0$' '776,0$' '665,0$' '526,0$' '1034,0$' '1403,0$'
 '1499,0$' '580,0$' '859,0$' '263,0$' '1376,0$' '885,0$' '1003,0$'
 '1321